src/c/coap-server.c

```diff
@@ -0,0 +1,465 @@
+/* CoAP server for device-coap-c
+ *
+ * Copyright (C) 2018 Olaf Bergmann <bergmann@tzi.org>
+ * Copyright (c) 2020 Ken Bannister
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <float.h>
+#include <signal.h>
+#include <string.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <stdbool.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <netdb.h>
+
+#include <coap2/coap.h>
+#include "edgex/devices.h"
+#include "device-coap.h"
+
+/* Maximum length of a string containing numeric values. */
+#define INT32_STR_MAXLEN 10
+#define FLOAT64_STR_MAXLEN (DBL_MAX_10_EXP + 2)
+
+#define RESOURCE_SEG1 "a1r"
+#define MSG_PAYLOAD_INVALID "payload not valid"
+#define MEDIATYPE_TEXT_PLAIN "text/plain"
+#define MEDIATYPE_APP_JSON "application/json"
+#define CONTENT_FORMAT_UNDEFINED UINT16_MAX
+
+static coap_driver *sdk_ctx;
+
+/* controls input loop */
+static int quit = 0;
+
+/* signal handler for input loop */
+static void
+handle_sig (int signum)
+{
+  (void)signum;
+  quit = 1;
+}
+
+/* Builds libcoap address struct from host/port */
+static int
+resolve_address (const char *host, const char *service, coap_address_t *lib_addr)
+{
+  struct addrinfo *res, *ainfo;
+  struct addrinfo hints;
+  int error, len=-1;
+
+  memset (&hints, 0, sizeof (hints));
+  memset (lib_addr, 0, sizeof (*lib_addr));
+  hints.ai_socktype = SOCK_DGRAM;
+  hints.ai_family = AF_UNSPEC;
+
+  error = getaddrinfo (host, service, &hints, &res);
+
+  if (error != 0)
+  {
+    iot_log_info (sdk_ctx->lc, "getaddrinfo: %s\n", gai_strerror (error));
+    return error;
+  }
+
+  for (ainfo = res; ainfo != NULL; ainfo = ainfo->ai_next)
+  {
+    len = lib_addr->size = ainfo->ai_addrlen;
```

Comment:
`len` is initialized to -1, and this value becomes the return value if `getaddrinfo()` doesn't resolve an Internet domain entry. A negative return value is handled in `run_server()` by logging an error and program exit. Am I missing something?

Reply:
I may be overthinking it, but could getaddrinfo theoretically return a number of entries none of which are AF_INET or AF_INET6?

Reply:
I see your point now. Yes, other address families will be rejected. Presently at the project level, use of IP families is a requirement because the port is hard coded to 5684 (PSK) or 5683 for NoSec.

Certainly we can leave the requirement to use IP in place for now, yet make `resolve_address()` more generic to be ready for more families in the future. Please confirm.

Reply:
ok

Reply:
Addressed in b28b46d